Add optional callbacks to action client for goal response and result

jacobperron opened the pull request:
Resolves #635.
This also makes it easier to incorporate action clients in composable nodes since we don't have to rely on waiting on futures.

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6726)](http://ci.ros2.org/job/ci_linux/6726/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3040)](http://ci.ros2.org/job/ci_linux-aarch64/3040/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5529)](http://ci.ros2.org/job/ci_osx/5529/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6556)](http://ci.ros2.org/job/ci_windows/6556/)

jacobperron commented:
Test failures appear unrelated. This is ready for review.

hidmic commented on rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp:
```diff
@@ -71,6 +73,9 @@ ClientGoalHandle<ActionT>::set_result(const WrappedResult & wrapped_result)
   std::lock_guard<std::mutex> guard(handle_mutex_);
   status_ = static_cast<int8_t>(wrapped_result.code);
   result_promise_.set_value(wrapped_result);
+  if (nullptr != result_callback_) {
```
@jacobperron nit: why not
```suggestion
  if (result_callback_) {
```
like it's done for the `goal_response_callback` above?

jacobperron replied:
eeed6f5

hidmic commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -318,31 +325,47 @@ class Client : public ClientBase
           } catch (...) {
             promise->set_exception(std::current_exception());
+            if (goal_response_callback) {
+              goal_response_callback(nullptr);
```
@jacobperron (meta, can ignore) what the callback gets as its sole argument isn't quite the same as what a wait on the future/promise would get. I wonder if we should be passing the resolved future/promise instead.

wjwwood replied:
I have to admit I don't understand what the semantic of passing nullptr here is. What scenario does this occur in?

jacobperron replied:
I believe this is the scenario where the "send result" request fails unexpectedly. In retrospect, I'm not sure what the correct behavior is in this case. The goal has already been accepted by the server, but we failed to request the result. Maybe we should attempt to cancel the goal? Maybe we actually return a goal handle, but throw an error if they attempt to get the result?

wjwwood replied:
I'm not sure what the right behavior is either, but it should probably be part of the documented behavior of these callbacks.

jacobperron replied:
I've updated the signature of the response callback to take a future (20757b4). Now, in this case calling `future::get()` will throw the exception.

hidmic commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -307,7 +310,11 @@ class Client : public ClientBase
+    GoalResponseCallback goal_response_callback = nullptr,
+    FeedbackCallback feedback_callback = nullptr,
+    ResultCallback result_callback = nullptr,
+    bool ignore_result = false)
```
@jacobperron (meta, can ignore): I know this is a purely subjective observation, biased by own prejudices, but since C++ lacks keyword arguments, I find having to pass multiple `nullptr` values to get to the one I actually care about somewhat annoying. Can we have some overloads for the most common use cases e.g. goal and feedback cb, goal, feedback cb and result cb, etc.?

wjwwood replied:
We've been going the way of "options structs" which alleviates some of this without expanding the number of signatures, also the signatures get confusing/ambiguous at some point.

But in general I agree that this becomes too fragile and inconvenient at some point.

jacobperron replied:
I could be persuaded to update this signature to accept a `SendGoalOptions` struct.

I was also thinking we might consider removing the `ignore_result` parameter and instead rely on checking if `result_callback` is set.

wjwwood replied:
Which ever is fine by me.

jacobperron replied:
I've added an options struct in f1c46f5. PTAL.

wjwwood replied:
Looks reasonable to me.

wjwwood commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -318,31 +325,47 @@ class Client : public ClientBase
+      feedback_callback,
+      result_callback,
+      ignore_result,
+      promise](
```
nitpick: this is a strange indention...

I would have done it like this:

```
[
  this,
  ...,
](...) mutable
{
  // ...
}
```

jacobperron replied:
eeed6f5

wjwwood commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -331,8 +331,7 @@ class Client : public ClientBase
       ignore_result,
-      promise](
-        std::shared_ptr<void> response) mutable
+      promise](std::shared_ptr<void> response) mutable
```
This wasn't what I was talking about on the style, the weird thing is that `[this` and `goal_request` use the same indention.

wjwwood replied:
Sorry, ignore this, I didn't see that you updated the signature again in the next commit.

jacobperron replied:
Right, FWIW, I recall trying to indent these lines as you suggested but one of the linters was complaining. Maybe because I didn't drop the closing `]` to the next line (I can't remember).

wjwwood commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -283,8 +283,11 @@ class Client : public ClientBase
+     * If the goal is accepted, then the pointer points to a valid goal handle.
+     * If the goal is rejected, then pointer has the value `nullptr`.
+     * If an error occurs while waiting for the goal response an exception will be thrown
+     * when calling `future::get()`.
```
It would be nice to list the at least some of the exceptions we know it can throw here.

jacobperron replied:
Done b69a1fd

jacobperron commented:
@wjwwood If there is an exception it is set here: 

https://github.com/ros2/rclcpp/blob/20757b4571312609e3e04f9232e70650c4742c40/rclcpp_action/include/rclcpp_action/client.hpp#L374

jacobperron commented:
@hidmic @wjwwood In addition to addressing your feedback (thanks!), I've added optional callback parameters to `async_get_result` and the cancel methods. I plan to squash these into three separate commits before merging. Ready for another round of review.

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6776)](http://ci.ros2.org/job/ci_linux/6776/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3077)](http://ci.ros2.org/job/ci_linux-aarch64/3077/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5572)](http://ci.ros2.org/job/ci_osx/5572/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6604)](http://ci.ros2.org/job/ci_windows/6604/)

CI also includes connected PRs:

- https://github.com/ros2/examples/pull/230
- https://github.com/ros2/system_tests/pull/346

hidmic commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -318,31 +352,38 @@ class Client : public ClientBase
-        if (!ignore_result) {
+        std::shared_ptr<GoalHandle> goal_handle(
+          new GoalHandle(goal_info, options.feedback_callback, options.result_callback));
+        if (options.result_callback) {
```
@jacobperron just for the record, this now means that if you just want the result future, you need to ask for it explicitly later on, always.

jacobperron replied:
Yeah, that's true. Though, this is a subtle feature. I bet that most use cases would provide a callback anyways and the case for sending a request early only to get a reference to the result future later is rare. The user can still get the future later with the client method, `async_get_result()`.

This brings me to another point. I think it's odd to have two public facing methods for the getting the result, `Client.async_get_result()` and `ClientGoalHandle.async_result()`. Originally, the client method was essentially bypassing the "ignore result" flag. I propose we make the goal handle method private and rely on the client method to return the future. Regarding the "ignore result" flag, in the case where the goal handle is not result aware, we have 2.5 options:

1. Keep the current behavior and make the goal handle result aware. I.e. Does not fail if the user previously set the "ignore result" flag.
    1. Or we can not even provide the "ignore result" option and make result requests on demand (this PR). 
2. Throw an exception, because the user chose to ignore the result explicitly.

Let me know what you think. 

To not drag this PR on any longer, I suggest we make any changes related to this discussion in a follow-up PR.

jacobperron commented on rclcpp_action/test/test_client.cpp:
```diff
@@ -344,14 +381,83 @@ TEST_F(TestClient, async_send_goal_with_feedback_and_result)
+  EXPECT_EQ(rclcpp_action::GoalStatus::STATUS_ACCEPTED, goal_handle->get_status());
+  EXPECT_FALSE(goal_handle->is_feedback_aware());
+  EXPECT_TRUE(goal_handle->is_result_aware());
+  auto future_result = action_client->async_get_result(goal_handle);
```
See my comment below proposing we make `goal_handle->async_result()` private.

ipa-mdl commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -318,31 +352,38 @@ class Client : public ClientBase
             this->make_result_aware(goal_handle);
           } catch (...) {
             promise->set_exception(std::current_exception());
+            options.goal_response_callback(future);
```
this does not check `goal_response_callback` first

jacobperron replied:
Good catch. I've swapped the two if-blocks as suggested in https://github.com/ros2/rclcpp/pull/738.

ipa-mdl commented on rclcpp_action/include/rclcpp_action/client.hpp:
```diff
@@ -318,31 +352,38 @@ class Client : public ClientBase
         goal_handles_[goal_handle->get_goal_id()] = goal_handle;
         promise->set_value(goal_handle);
+        if (options.goal_response_callback) {
+          options.goal_response_callback(future);
```
This might get called after `result_callback` (not 100% sure)
